build.rs

```diff
@@ -1,14 +1,109 @@
 extern crate cbindgen;
 
-#[cfg(feature = "cheader")]
 use std::env;
-#[cfg(feature = "cheader")]
+use std::ffi::OsStr;
+use std::ops::Deref as _;
 use std::path::Path;
+use std::process::Command;
+use std::process::Stdio;
+use std::str;
+
+use anyhow::bail;
+use anyhow::Context as _;
+use anyhow::Result;
+
+/// Format a command with the given list of arguments as a string.
+fn format_command<C, A, S>(command: C, args: A) -> String
+where
+    C: AsRef<OsStr>,
+    A: IntoIterator<Item = S>,
+    S: AsRef<OsStr>,
+{
+    args.into_iter().fold(
+        command.as_ref().to_string_lossy().into_owned(),
+        |mut cmd, arg| {
+            cmd += " ";
+            cmd += arg.as_ref().to_string_lossy().deref();
+            cmd
+        },
+    )
+}
+
+/// Run a command with the provided arguments.
+fn run<C, A, S>(command: C, args: A) -> Result<()>
+where
+    C: AsRef<OsStr>,
+    A: IntoIterator<Item = S> + Clone,
+    S: AsRef<OsStr>,
+{
+    let mut instance = Command::new(command.as_ref())
+        .stdin(Stdio::null())
+        .stdout(Stdio::null())
+        .args(args.clone())
+        .output()
+        .with_context(|| {
+            format!(
+                "failed to run `{}`",
+                format_command(command.as_ref(), args.clone())
+            )
+        })?;
+
+    if !instance.status.success() {
+        let code = if let Some(code) = instance.status.code() {
+            format!(" ({})", code)
+        } else {
+            String::new()
+        };
+
+        let _last = instance.stderr.pop();
+        let stderr = if let Ok(stderr) = str::from_utf8(&instance.stderr) {
+            format!(": {stderr}")
+        } else {
+            format!(": {:?}", &instance.stderr)
+        };
+
+        bail!(
+            "`{}` reported non-zero exit-status{code}{stderr}",
+            format_command(command, args),
+        );
+    }
+
+    Ok(())
+}
+
+/// Build `data/test.bin`
+fn build_dwarf_v4_test_bin(test_bin: &Path) {
+    let mut output = test_bin.to_path_buf();
+    assert!(output.set_extension("bin"));
+
+    // Ideally we'd use the `cc` crate here, but it seemingly can't be convinced
+    // to create binaries.
+    run(
+        "cc",
+        [
+            "-gdwarf-4".as_ref(),
+            test_bin.as_os_str(),
+            "-o".as_ref(),
+            output.as_os_str(),
+        ],
+    )
+    .expect("failed to run `cc`")
+}
+
+/// Build the various test binaries.
+fn build_test_bins(crate_root: &Path) {
+    let path = crate_root.join("data").join("test.c");
+    println!("cargo:rerun-if-changed={}", path.display());
+    build_dwarf_v4_test_bin(&path);
+}
 
 fn main() {
+    let crate_dir = env!("CARGO_MANIFEST_DIR");
+
+    build_test_bins(crate_dir.as_ref());
```

Comment:
The build script is a separate binary and unfortunately does not get the same information as a the main program or library would from what I understand. So the only choices I saw were:
- build unconditionally (what I did)
- introduce a feature for building the binaries and then fail the tests if binaries are not present because feature is not enabled (seems clunky and unnecessarily at this stage)
- introduce a feature for rebuilding the binaries and then check the binary in (seems unnecessary at this stage; didn't want to do that because it's not necessary and not good practice to check binaries in)
- add a bunch of `make` hocus pocus that builds binaries and then runs tests (not great either, in my opinion)

So pick your poison. I think the best option may be to have a feature to opt out of building. This way, if no suitable compiler is available (or whatever other reason there may be) it could be set to not build this stuff.

Reply:
> So pick your poison. I think the best option may be to have a feature to opt out of building. This way, if no suitable compiler is available (or whatever other reason there may be) it could be set to not build this stuff.

How do you think to let the case call a function to check the binary and build it if necessary?

Reply:
> How do you think to let the test case call a function to check the binary and build it if necessary?

It's reasonably for a binary to be used by multiple tests (and already been done). So then we'd need to synchronize builds. Now we'd also get into the business of defining what "necessary" means: conditional rebuilding based on time stamps or changes to source files etc. I don't want to go down that road at this point and it does not change anything substantially: if we want to fend off the non-availability of development executables (e.g., because a developer does not care about this part and does not have tooling installed), we'd still want features.
So I think opting out of rebuilding via feature is the best approach right now. Later (should we require more fringe tools that we can't justify everybody installing) we can invert the logic to an opt-in based approach. And of course we can always split out features then: e.g., `test-gym` would build and test with relevant `gsym` binaries and require the requisite tooling, while `test-dwarf` would take care of the same for DWARF logic.

Does that sound okay?